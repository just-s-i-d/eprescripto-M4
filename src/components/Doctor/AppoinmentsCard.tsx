import { useState } from "react";
import { Card, Skeleton, Avatar, Tooltip, Image } from "antd";
import { CloseCircleTwoTone, PhoneTwoTone } from "@ant-design/icons";

import noDataPng from "@assets/no-data.png";
import CardTitle from "@components/ui/CardTitle";
import PopModal from "@components/ui/PopModal";
import profilePic from "@assets/profilepic.png";
import {
  ApiResponseData,
  ApiResponseDataType,
  AppointmentDataType,
} from "@constants/types";
import { cancelAppointment } from "@utils/Doctor";
import ErrorBoundary from "@components/ErrorBoundary";
import useStatesHook from "src/hooks/useStatesHook";
import { showToast } from "@utils/common";

type AppointmentCardPropsType = {
  appointments?: ApiResponseData<AppointmentDataType>;
  loading: boolean;
  appointmentsState: ReturnType<
    typeof useStatesHook<ApiResponseData<AppointmentDataType>>
  >;
  getDataForAppointments: () => void;
};

const AppoinmentsCard = ({
  appointments,
  loading,
  appointmentsState,
  getDataForAppointments,
}: AppointmentCardPropsType) => {
  const [open, setOpen] = useState<boolean>(false);
  const [selectedPatient, setSelectedPatient] =
    useState<ApiResponseDataType<AppointmentDataType>>();
  const handlerCancleAppointment = (
    patient: ApiResponseDataType<AppointmentDataType>,
  ) => {
    setOpen(true);
    setSelectedPatient(patient);
  };
  const confirmHandler = async () => {
    if (selectedPatient) {
      cancelAppointment(selectedPatient)
        .then((res) => {
          showToast(res.message, res.type);
          getDataForAppointments();
        })
        .catch((error) => {
          showToast(error.message, error.type);
        });
    }
  };
  return (
    <>
      <Card
        className={`w-[32%] max-xxl:h-[59vh] max-h-[59vh] max-xl:max-h-[60vh] max-xxl:w-[48%] max-xl:w-[48%] max-md:w-9/12 max-sm:w-full `}
        bordered={false}
      >
        <ErrorBoundary
          refreshComponent={() => appointmentsState.setRefresh((prev) => !prev)}
          error={appointmentsState.error}
        >
          <CardTitle>Today's Appointments</CardTitle>
          {!appointmentsState.loading && !appointmentsState.data?.length ? (
            <div className="flex flex-col items-center justify-center">
              <Image src={noDataPng} preview={false} />
              <h2 className="text-lg mt-12 text-center">No Appointments</h2>
            </div>
          ) : (
            <div className="mx-auto px-4 h-[44vh] overflow-y-auto">
              <ul className="mt-1 divide-y">
                {appointments?.map((item, idx) => (
                  <li
                    key={idx}
                    className="py-4 mb-2 px-2 bg-[#F0F3F4] rounded-lg"
                  >
                    <div className="flex items-start justify-between">
                      <div className="flex gap-3">
                        {loading ? (
                          <Skeleton.Avatar active size={48} />
                        ) : (
                          <Avatar
                            size={48}
                            src={item.attributes?.profilePic || profilePic}
                          />
                        )}
                        <Skeleton
                          loading={loading}
                          active
                          avatar={false}
                          title={false}
                          paragraph={{ rows: 2, width: [100, 100] }}
                        >
                          <div>
                            <span className="block text-base text-gray-700">
                              {item.attributes?.pName}
                            </span>
                            <span className="block text-[12px] text-gray-600">
                              {item.attributes?.timeSlot}
                            </span>
                          </div>
                        </Skeleton>
                      </div>
                      {loading ? (
                        <Skeleton.Button active />
                      ) : (
                        <span
                          className={`py-1 px-2 rounded-lg ${item.attributes?.status.toLowerCase()}`}
                        >
                          {item.attributes?.status}
                        </span>
                      )}
                    </div>
                    <div className="flex justify-between">
                      <span className="block text-sm text-gray-700 font-bold">
                        <span className="mr-1 py-1 text-gray-500 font-semibold">
                          Referrer
                        </span>
                        {item.attributes.referrer || "NA"}
                      </span>
                      <div className="flex justify-end gap-4 text-lg">
                        <Tooltip title={item.attributes?.contact}>
                          <PhoneTwoTone
                            className="rotate-90 cursor-pointer"
                            href={`tel:${item.attributes?.contact}`}
                          />
                        </Tooltip>

                        {item.attributes.status === "Pending" && (
                          <CloseCircleTwoTone
                            className="cursor-pointer"
                            onClick={() => handlerCancleAppointment(item)}
                          />
                        )}
                      </div>
                      <PopModal
                        setOpen={setOpen}
                        open={open}
                        okButtonText="Sure"
                        title="Cancel Appointment"
                        footer={true}
                        confirmHandler={confirmHandler}
                      >
                        Are you sure,you want to cancel the appoinment for{" "}
                        {selectedPatient?.attributes.pName}?
                      </PopModal>
                    </div>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </ErrorBoundary>
      </Card>
    </>
  );
};

export default AppoinmentsCard;
