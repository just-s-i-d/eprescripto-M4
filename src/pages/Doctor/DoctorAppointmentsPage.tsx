import { useEffect, useState } from "react";

import { ColumnsType } from "antd/es/table";
import { Card, Tooltip } from "antd";
import { PhoneTwoTone, CloseCircleTwoTone } from "@ant-design/icons";

import ErrorBoundary from "@components/ErrorBoundary";
import useStatesHook from "../../hooks/useStatesHook";

import {
  cancelAppointment,
  formatDateReadable,
  getDoctorAppointmentsData,
} from "@utils/Doctor";
import {
  AppointmentDataType,
  DoctorAppointmentsDataType,
} from "@constants/types";
import TableCard from "@components/ui/TableCard";
import PopModal from "@components/ui/PopModal";
import { showToast } from "@utils/common";

const DoctorAppointmentsPage = () => {
  const appointmentsTable = useStatesHook<DoctorAppointmentsDataType>();
  const [open, setOpen] = useState(false);
  const [selectedPatient, setSelectedPatient] = useState<AppointmentDataType>();
  const handlerCancleAppointment = (patient: AppointmentDataType) => {
    setOpen(true);
    setSelectedPatient(patient);
  };
  const confirmHandler = async () => {
    if (selectedPatient) {
      const response = await cancelAppointment(selectedPatient);
      response && showToast(response.message, response.type);
    }
  };
  useEffect(() => {
    getDoctorAppointmentsData()
      .then((res) => {
        appointmentsTable.setData(res);
        appointmentsTable.setError(false);
        appointmentsTable.setLoading(false);
      })
      .catch(() => {
        appointmentsTable.setError(true);
        appointmentsTable.setLoading(false);
      });
  }, []);
  const doctorAppointmentsTableColumns: ColumnsType<AppointmentDataType> = [
    {
      title: "Patient Id",
      dataIndex: "pId",
      width: 50,
    },
    {
      title: "Patient Name",
      dataIndex: "pName",
      width: 150,
    },
    {
      title: "Email",
      dataIndex: "email",
      width: 200,
    },
    {
      title: "Date",
      dataIndex: "date",
      width: 150,
      render: (date, record) => (
        <div className="flex flex-col">
          <span>{formatDateReadable(date)}</span>
          <span>{record.timeSlot}</span>
        </div>
      ),
    },
    {
      title: "Referrer",
      dataIndex: "referer",
      width: 100,
    },
    {
      title: "Status",
      dataIndex: "status",
      width: 50,
      render: (value: string) => (
        <span className={`py-1 px-2 rounded-lg ${value.toLowerCase()}`}>
          {value}
        </span>
      ),
      filters: [
        { text: "Pending", value: "Pending" },
        { text: "Attended", value: "Attended" },
        { text: "Cancelled", value: "Cancelled" },
      ],
      onFilter: (value, record) => record.status === value,
    },
    {
      title: "Action",
      width: 100,
      render: (_, record) => (
        <div className="flex gap-4 text-lg">
          <Tooltip title={record.contact}>
            <PhoneTwoTone
              className="rotate-90 cursor-pointer fill-[#19504c]"
              href={`tel:${record.contact}`}
            />
          </Tooltip>
          {record.status === "Pending" && (
            <CloseCircleTwoTone
              className="cursor-pointer"
              onClick={() => handlerCancleAppointment(record)}
            />
          )}
        </div>
      ),
    },
  ];
  return (
    <>
      <ErrorBoundary>
        <Card className="appointment-table-card h-min w-full overflow-x-scroll">
          <TableCard
            className="min-w-[1100px]"
            columns={doctorAppointmentsTableColumns}
            tableData={appointmentsTable.data}
            pageSize={8}
          />
          <PopModal
            setOpen={setOpen}
            open={open}
            okButtonText="Sure"
            title="Cancel Appointment"
            footer={true}
            confirmHandler={confirmHandler}
          >
            Are you sure,you want to cancel the appoinment for{" "}
            {selectedPatient?.pName}
          </PopModal>
        </Card>
      </ErrorBoundary>
    </>
  );
};
export default DoctorAppointmentsPage;
