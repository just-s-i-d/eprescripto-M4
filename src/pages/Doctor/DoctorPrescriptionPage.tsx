import MedicationForm from "@components/Doctor/MedicationForm";
import NewPatientCard from "@components/Doctor/NewPatientCard";
import ErrorBoundary from "@components/ErrorBoundary";
import CardTitle from "@components/ui/CardTitle";
import PopModal from "@components/ui/PopModal";
import {
  ApiResponseData,
  PatientDataType,
  PatientInfoType,
  SelectOptionsType,
} from "@constants/types";
import {
  addNewPatient,
  createNewPrescription,
  getData,
  patientsEndpoint,
} from "@utils/Doctor";
import { showToast } from "@utils/common";
import { Button, Card, Form, Input, Select } from "antd";
import { useEffect, useState } from "react";
import { MedicationType } from "../../constants/types";
import {
  createOptions,
  medicationInputFieldGenerator,
} from "../../utils/Doctor";
import { SELECT_GENDER_OPTIONS } from "@constants/constants";
const { useForm, useWatch } = Form;
const { TextArea } = Input;
const DoctorPrescriptionPage = () => {
  const [open, setOpen] = useState(false);
  const [showAllergies, setShowAllergies] = useState(false);
  const [disabled, setDisabled] = useState(false);
  const [newPatientForm] = useForm();
  const [prescriptionForm] = useForm();
  const formValues = useWatch([], newPatientForm);
  const [patientsData, setPatientsData] = useState<PatientInfoType[]>();
  const [selectedPatient, setSelectedPatient] = useState<PatientInfoType>();
  const [options, setOptions] = useState<SelectOptionsType>();
  const prescriptionFormValues = useWatch([], prescriptionForm);
  const [medicationInputFields, setMedicationInputFiels] = useState<
    MedicationType[]
  >([medicationInputFieldGenerator()]);
  const addMedicationInput = () => {
    const inputField = medicationInputFieldGenerator();
    setMedicationInputFiels((prev) => [...prev, inputField]);
  };
  const removeMedicationInput = (id: number) => {
    const reducedInputFields = medicationInputFields.filter(
      (item) => item.id !== id,
    );
    setMedicationInputFiels(reducedInputFields);
  };
  const updatedData = () => {
    getData<ApiResponseData<PatientDataType>>(patientsEndpoint).then((res) => {
      const data = res.map((element) => ({
        id: element.id,
        ...element.attributes,
      }));
      setPatientsData(data);
      setOptions(createOptions(data));
    });
  };
  const onChangeHanlder = (id: number) => {
    if (patientsData) {
      const patient = patientsData.find(
        (patient: PatientInfoType) => patient.id === id,
      );
      setSelectedPatient(patient);
    }
  };
  const submitHanlder = () => {
    const patientInfo = formValues;
    addNewPatient(patientInfo)
      .then(() => {
        showToast("New patient added", "success");
        updatedData();
      })
      .catch((res) => {
        showToast(res.message, res.type);
      });
    newPatientForm.resetFields();
  };
  const cancelHanlder = () => {
    prescriptionForm.resetFields();
    setSelectedPatient(undefined);
  };
  const prescriptionFormSubmitHandler = () => {
    createNewPrescription(prescriptionFormValues)
      .then((res) => {
        showToast(res.message, res.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      })
      .finally(() => {
        prescriptionForm.resetFields();
        setSelectedPatient(undefined);
      });
  };
  useEffect(() => {
    updatedData();
  }, []);
  useEffect(() => {
    prescriptionForm.resetFields();
  }, [selectedPatient]);
  useEffect(() => {
    prescriptionForm.validateFields({ validateOnly: true }).then(
      () => {
        setDisabled(false);
      },
      () => {
        setDisabled(true);
      },
    );
  }, [prescriptionFormValues]);

  return (
    <ErrorBoundary>
      <Card>
        <CardTitle className="flex justify-between">
          Prescription Form{" "}
          <Button
            className="text-white bg-secondary"
            onClick={() => setOpen(true)}
            data-testid="add-patient-btn"
          >
            New Patient
          </Button>
        </CardTitle>
        <Form
          data-testid="prescription-form"
          form={prescriptionForm}
          layout="vertical"
          className="custom-form custom-form-profile"
          initialValues={selectedPatient}
        >
          <div className="grid sm:grid-cols-2 md:grid-cols-3 xl:grid-cols-4 gap-x-6 gap-y-3">
            <Form.Item name="pId" label="Patient" className="w-full" required>
              <Select
                data-testid="patient-select"
                placeholder="Select a Patient"
                options={options}
                onChange={onChangeHanlder}
                value={selectedPatient?.id}
              />
            </Form.Item>
            <Form.Item
              className="w-full "
              name="pName"
              label="Patient Name"
              required
            >
              <Input disabled />
            </Form.Item>
            <Form.Item
              className="w-full"
              name="dob"
              label="Date of Birth"
              required
            >
              <Input type="date" disabled />
            </Form.Item>
            <Form.Item
              className="xs:w-full"
              name="gender"
              label="Gender"
              required
              rules={[
                {
                  required: true,
                  message: "Field cannot be empty",
                },
              ]}
            >
              <Select
                disabled
                placeholder="Select a gender"
                options={SELECT_GENDER_OPTIONS}
              />
            </Form.Item>
            <Form.Item
              name="prescriptionName"
              label="Prescription Name"
              className="xs:w-full"
              required
              rules={[
                {
                  required: true,
                  message: "Field cannot be empty",
                  validator: (_, value) => {
                    if (value?.trim()) return Promise.resolve();
                    else return Promise.reject();
                  },
                },
                {
                  pattern: /^[A-Za-z0-9\s-]+$/,
                  message: "Special character not allowed",
                },
                {
                  max: 20,
                  message: "Only 20 characters allowed",
                },
              ]}
            >
              <Input />
            </Form.Item>
            <Form.Item
              className="xs:w-full"
              name="nextVisit"
              label="Next Visit"
              rules={[{ pattern: /^[0-9]+$/, message: "Enter valid integer" }]}
            >
              <Input type="number" placeholder="In Days" />
            </Form.Item>
            <Form.Item name="notes" label="Important Notes" className="w-full">
              <TextArea rows={1} />
            </Form.Item>
          </div>
          <div className="w-full mt-4">
            {medicationInputFields.map((item, index) => (
              <MedicationForm
                key={item.id}
                fields={item}
                removeMedicationInput={removeMedicationInput}
                removeable={index > 0}
              />
            ))}
            <div className="flex justify-between mt-2 mb-8">
              <Button onClick={addMedicationInput}>Add medication</Button>
              <Button onClick={() => setShowAllergies(true)}>
                Show Patient Allergies
              </Button>
            </div>
          </div>
          <div className="flex gap-4">
            <Button
              onClick={prescriptionFormSubmitHandler}
              disabled={disabled}
              className="text-white bg-secondary disabled:opacity-50"
            >
              Create Prescription
            </Button>
            <Button onClick={cancelHanlder}>Cancel</Button>
          </div>
        </Form>
      </Card>
      <ErrorBoundary>
        <PopModal
          open={open}
          setOpen={setOpen}
          title="New Patient"
          okButtonText="Add Patient"
          confirmHandler={submitHanlder}
          footer={true}
        >
          <NewPatientCard newPatientForm={newPatientForm} />
        </PopModal>
      </ErrorBoundary>
      <ErrorBoundary>
        <PopModal
          footer={false}
          open={showAllergies}
          setOpen={setShowAllergies}
          title="Patient Allergies"
        >
          <div className="min-h-[10vh] max-h-[20vh] mt-4 text-lg overflow-scroll">
            {selectedPatient?.allergies ? (
              <>
                <div>Patient has following allergies :</div>
                <div>{selectedPatient?.allergies}</div>
              </>
            ) : (
              <div>Patient has no allergies mentioned</div>
            )}
          </div>
        </PopModal>
      </ErrorBoundary>
    </ErrorBoundary>
  );
};
export default DoctorPrescriptionPage;
