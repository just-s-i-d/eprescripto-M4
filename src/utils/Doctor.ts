import axios from "axios";

import { getLabels } from "@constants/constants";
import {
  AppointmentDataType,
  AppointmentsDataType,
  ChartApiDataType,
  DoctorAppointmentsDataType,
  InfoCardDetailsType,
  LineChartApiResDataType,
  LineChartDataType,
  PrescrtionsDataType,
  ReviewsDataType,
} from "@constants/types";

//endpoints for data
export const infoCardDataEndPoint =
  "https://api.npoint.io/bd6151d98738374a2236";

export const appointmentEndPoint = "https://api.npoint.io/9cc072a9f2c41c77a40a";

export const prescriptionsEndPoint =
  "https://api.npoint.io/7c31657f364397ae7175";

export const pieChartDataEndPoint =
  "https://api.npoint.io/f8eeff1048038038110e";

export const lineChartDataEndPoint =
  "https://api.npoint.io/75d76a4af2cc563e20a8";

export const barChartDataEndPoint =
  "https://api.npoint.io/bb362ddf298c94c14975";

export const doctorAppointmentsEndPoint =
  "https://api.npoint.io/a7f8fadf08c1b182fc98";

export const doctorReviewsEndPoint =
  "https://api.npoint.io/d21b0ee032131e268c76";

export function formatDateReadable(dateString: string) {
  const options = { year: "numeric", month: "long", day: "numeric" };
  const formattedDate = new Date(dateString).toLocaleDateString(
    undefined,
    options,
  );
  return formattedDate;
}
export const getInfoCardsData = async (): Promise<InfoCardDetailsType> => {
  return new Promise((resolve, reject) => {
    try {
      axios
        .get(infoCardDataEndPoint)
        .then((res) => {
          resolve(res.data.infoCardsData);
        })
        .catch(() => {
          reject("Cannot get data");
        });
    } catch {
      reject("Server error");
    }
  });
};

export const getAppointmentData = async (): Promise<AppointmentsDataType> => {
  return new Promise((resolve, reject) => {
    try {
      axios
        .get(appointmentEndPoint)
        .then((res) => {
          resolve(res.data.appointments);
        })
        .catch(() => {
          reject("Cannot get data");
        });
    } catch {
      reject("Server error");
    }
  });
};

export const getChartData = async (
  endPoint: string,
): Promise<ChartApiDataType> => {
  return new Promise((resolve, reject) => {
    try {
      axios
        .get(endPoint)
        .then((res) => {
          resolve(res.data);
        })
        .catch(() => {
          reject("Cannot get data");
        });
    } catch (error) {
      reject("Server Connection Error");
    }
  });
};

export const getPrescriptions = async (): Promise<PrescrtionsDataType> => {
  return new Promise((resolve, reject) => {
    try {
      axios
        .get(prescriptionsEndPoint)
        .then((res) => {
          resolve(res.data.prescriptions);
        })
        .catch(() => {
          reject("Cannot get data");
        });
    } catch (error) {
      reject("Server Connection Error");
    }
  });
};

export const getDataForLineGraph = (
  data: LineChartApiResDataType,
  days: number = 7,
): LineChartDataType => {
  const labels = getLabels(days);
  switch (days) {
    case 7:
      return { labels, data: data.seven };
    case 10:
      return {
        labels,
        data: data.ten,
      };
    case 30:
      return {
        labels,
        data: data.thirty,
      };
    default:
      return { labels, data: data.seven };
  }
};

export const cancelAppointment = async (appointment: AppointmentDataType) => {
  try {
    const { pId } = appointment;
    const res = await axios.get(appointmentEndPoint);
    if (res.status >= 200 && res.status <= 299) {
      const appointments = res.data.appointments;
      appointments.filter(
        (patient: AppointmentDataType) => patient.pId !== pId,
      );
      return { message: "Appointment Cancelled", type: "success" };
    } else {
      return { message: "Server error", type: "error" };
    }
  } catch (error) {
    return { message: "Server Connection Error", type: "error" };
  }
};

export const getDoctorAppointmentsData =
  async (): Promise<DoctorAppointmentsDataType> => {
    return new Promise((resolve, reject) => {
      try {
        axios
          .get(doctorAppointmentsEndPoint)
          .then((res) => {
            resolve(res.data.appointments);
          })
          .catch(() => {
            reject("Cannot get data");
          });
      } catch {
        reject("Server error");
      }
    });
  };

export const getReviewsData = async (): Promise<ReviewsDataType> => {
  return new Promise((resolve, reject) => {
    try {
      axios
        .get(doctorReviewsEndPoint)
        .then((res) => {
          resolve(res.data.reviews);
        })
        .catch(() => {
          reject("Cannot get data");
        });
    } catch {
      reject("Server error");
    }
  });
};
