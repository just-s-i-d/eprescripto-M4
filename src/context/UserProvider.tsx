import {
  PropsWithChildren,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from "react";

import { getUserData } from "@utils/Doctor";
import { showToast } from "@utils/common";

type UserContextType = {
  isLoggedIn?: boolean;
  setIsLoggedIn?: React.Dispatch<SetStateAction<boolean>>;
  isDarkTheme?: boolean;
  setIsDarkTheme?: React.Dispatch<SetStateAction<boolean>>;
  role?: string;
  setRole?: React.Dispatch<SetStateAction<string | undefined>>;
  refresh?: boolean;
  setRefresh?: React.Dispatch<SetStateAction<boolean | undefined>>;
  getRole?: () => void;
};

export const UserContext = createContext<UserContextType>({});
export const UserProvider = ({ children }: PropsWithChildren) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [isDarkTheme, setIsDarkTheme] = useState<boolean>(false);
  const [role, setRole] = useState<string>();

  const getRole = () => {
    getUserData()
      .then((res) => {
        setRole(res.role?.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      });
  };
  useEffect(() => {
    const token = localStorage.getItem("token");
    if (token) {
      getRole();
    }
    setIsLoggedIn(!!token);
  }, []);
  const value = {
    isLoggedIn,
    setIsLoggedIn,
    role,
    setRole,
    isDarkTheme,
    setIsDarkTheme,
    getRole,
  };
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};
