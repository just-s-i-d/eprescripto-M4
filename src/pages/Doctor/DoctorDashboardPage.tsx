import { useEffect } from "react";

import NextPatientCard from "@components/Doctor/NextPatientCard";
import InfoCards from "@components/Doctor/InfoCards";
import ErrorBoundary from "@components/ErrorBoundary";
import LineChart from "@components/Charts/LineChart";
import BarChart from "@components/Charts/BarChart";
import AppoinmentsCard from "@components/Doctor/AppoinmentsCard";
import DoughnutChart from "@components/Charts/DoughnutChart";
import {
  appointmentsEndPoint,
  barChartDataEndPoint,
  doughnutChartDataEndPoint,
  getData,
  getDataForLineGraph,
  infoCardDataEndPoint,
  lineChartDataEndPoint,
} from "@utils/Doctor";
import useStatesHook from "../../hooks/useStatesHook";
import {
  ApiResponseData,
  ApiResponseDataType,
  AppointmentDataType,
  ChartDataType,
  InfoCardDetailsType,
  LineChartApiResDataType,
} from "@constants/types";

const DoctorDashboardPage: React.FC = () => {
  const infoCards = useStatesHook<ApiResponseData<InfoCardDetailsType>>();
  const appointments = useStatesHook<ApiResponseData<AppointmentDataType>>();
  const doughnutChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
  const barChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
  const lineChart =
    useStatesHook<ApiResponseDataType<LineChartApiResDataType>>();
  useEffect(() => {
    getData(infoCardDataEndPoint)
      .then((res) => {
        infoCards.setData(res);
        infoCards.setError(false);
      })
      .catch(() => {
        infoCards.setError(true);
      });
  }, [infoCards.refresh]); //eslint-disable-line

  useEffect(() => {
    getData(appointmentsEndPoint)
      .then((res) => {
        appointments.setData(res);
        appointments.setLoading(false);
        appointments.setError(false);
      })
      .catch(() => {
        appointments.setError(true);
        appointments.setLoading(false);
      });
  }, [appointments.refresh]); //eslint-disable-line

  useEffect(() => {
    getData(doughnutChartDataEndPoint)
      .then((res) => {
        doughnutChart.setData(res);
        doughnutChart.setLoading(false);
        doughnutChart.setError(false);
      })
      .catch(() => {
        doughnutChart.setError(true);
      });
  }, [doughnutChart.refresh]); //eslint-disable-line

  useEffect(() => {
    getData(lineChartDataEndPoint)
      .then((res) => {
        lineChart.setData(res);
        lineChart.setError(false);
      })
      .catch(() => {
        lineChart.setError(true);
      });
  }, [lineChart.refresh]); //eslint-disable-line

  useEffect(() => {
    getData(barChartDataEndPoint)
      .then((res) => {
        barChart.setData(res);
        barChart.setError(false);
      })
      .catch(() => {
        barChart.setError(true);
      });
  }, [barChart.refresh]); //eslint-disable-line

  return (
    <>
      <ErrorBoundary>
        <div className="max-w-full flex">
          <div className="w-full flex flex-col gap-6">
            <ErrorBoundary>
              <InfoCards cardDetails={infoCards.data} infoCards={infoCards} />
            </ErrorBoundary>
            <div className="flex flex-wrap justify-between gap-y-6 max-tablet:justify-center">
              <ErrorBoundary>
                <NextPatientCard
                  appointments={appointments}
                  appointment={appointments.data && appointments.data[0]}
                  error={appointments.error}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <DoughnutChart
                  doughnutChart={doughnutChart}
                  chartData={doughnutChart.data}
                  title="Ratings Composition"
                  label="Reviews"
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <AppoinmentsCard
                  appointmentsState={appointments}
                  appointments={appointments.data}
                  loading={appointments.loading}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <LineChart
                  lineChart={lineChart}
                  chartData={lineChart.data}
                  title="Patients in Last 7 Days"
                  xAxesTitle="Dates"
                  yAxesTitle="Number of Patients"
                  getDataForLineGraph={getDataForLineGraph}
                />
              </ErrorBoundary>
              <ErrorBoundary>
                <BarChart
                  barChart={barChart}
                  chartData={barChart.data}
                  title="Age Groups Comparison"
                  label="Patients"
                  xAxesTitle="Different Age Groups (In Years)"
                  yAxesTitle="Number of Patients"
                />
              </ErrorBoundary>
            </div>
          </div>
        </div>
      </ErrorBoundary>
    </>
  );
};

export default DoctorDashboardPage;
