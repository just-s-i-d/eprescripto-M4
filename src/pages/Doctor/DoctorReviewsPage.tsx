import { useEffect } from "react";

import { Card, Table } from "antd";
import { ColumnsType } from "antd/es/table";

import { ApiResponseData, ReviewDataType } from "@constants/types";
import { getData, reviewsDataEndPoint } from "@utils/Doctor";
import useStatesHook from "../../hooks/useStatesHook";
import starPng from "@assets/star.png";
import ErrorBoundary from "@components/ErrorBoundary";

const DoctorReviewsPage = () => {
  const reviews = useStatesHook<Record<string, string | number>[]>();
  const columns: ColumnsType<ReviewDataType> = [
    {
      title: "S. No",
      render: (_, _1, index) => index + 1,
    },
    {
      title: "Rating",
      dataIndex: "rating",
      key: "rating",
      render: (rating: number) => (
        <span className="flex">
          {Array.from({ length: rating }, (_, index) => (
            <img
              src={starPng}
              key={index}
              role="img"
              aria-label="star"
              width={20}
            />
          ))}
        </span>
      ),
      filters: [
        {
          text: "1 Star",
          value: 1,
        },
        {
          text: "2 Star",
          value: 2,
        },
        {
          text: "3 Star",
          value: 3,
        },
        {
          text: "4 Star",
          value: 4,
        },
        {
          text: "5 Star",
          value: 5,
        },
      ],
      onFilter: (value, record) => record.rating === value,
      sorter: {
        compare: (a, b) => a.rating - b.rating,
        multiple: 3,
      },
    },
    {
      title: "Comment",
      dataIndex: "comment",
      key: "comment",
    },
  ];
  useEffect(() => {
    getData<ApiResponseData<Record<string, string | number>>>(
      reviewsDataEndPoint,
    )
      .then((res) => {
        const data = res.map((element) => element.attributes);
        reviews.setData(data);
        reviews.setError(false);
        reviews.setLoading(false);
      })
      .catch(() => {
        reviews.setError(true);
        reviews.setLoading(false);
      });
  }, [reviews.refresh]);
  return (
    <>
      <Card className="overflow-scroll min-h-[50vh]">
        <ErrorBoundary
          error={reviews.error}
          refreshComponent={() => reviews.setRefresh((prev) => !prev)}
        >
          <Table
            className="custom-table min-w-[800px]"
            columns={columns as ColumnsType<Record<string, string | number>>}
            dataSource={reviews.data}
            loading={reviews.loading}
          />
        </ErrorBoundary>
      </Card>
    </>
  );
};
export default DoctorReviewsPage;
