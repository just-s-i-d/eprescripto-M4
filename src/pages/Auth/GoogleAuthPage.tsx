import Loading from "@components/ui/Loading";
import { UserContext } from "@context/UserProvider";
import { getJwtToken } from "@utils/Doctor";
import { useContext, useEffect } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

const GoogleAuthPage = () => {
  const { setIsLoggedIn, getRole } = useContext(UserContext);
  const [params] = useSearchParams();
  const navigate = useNavigate();
  useEffect(() => {
    const idToken = params.get("id_token");
    const accessToken = params.get("access_token");
    if (idToken && accessToken) {
      getJwtToken(idToken, accessToken)
        .then(() => {
          setIsLoggedIn && setIsLoggedIn(true);
          getRole && getRole();
          navigate("/dashboard", { replace: true });
        })
        .catch(() => {
          navigate("/auth", { replace: true });
        });
    }
  }, []);
  return <Loading />;
};
export default GoogleAuthPage;
