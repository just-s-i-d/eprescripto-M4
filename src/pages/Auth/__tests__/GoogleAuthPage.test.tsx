import { render, waitFor } from "@testing-library/react";
import GoogleAuthPage from "../GoogleAuthPage";
import { MemoryRouter } from "react-router-dom";
import { getJwtToken } from "@utils/Doctor";
import { UserContext } from "@context/UserProvider";

jest.mock("@utils/Doctor", () => ({
  getJwtToken: jest.fn().mockResolvedValue("mocked result"),
}));

type WrapperComponentProps = {
  mockedSetIsloggedIn: VoidFunction;
  mockedGetRole: VoidFunction;
};
const WrapperComponent = ({
  mockedSetIsloggedIn,
  mockedGetRole,
}: WrapperComponentProps) => {
  return (
    <MemoryRouter
      initialEntries={[
        "/auth/google?id_token=mockedToken&access_token=mockedTken",
      ]}
    >
      <UserContext.Provider
        value={{ setIsLoggedIn: mockedSetIsloggedIn, getRole: mockedGetRole }}
      >
        <GoogleAuthPage />
      </UserContext.Provider>
    </MemoryRouter>
  );
};
describe("Google Auth page tests", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  const mockedSetIsloggedIn = jest.fn();
  const mockedGetRole = jest.fn();
  it("Getting Jwt token and role", async () => {
    const { asFragment } = render(
      <WrapperComponent
        mockedGetRole={mockedGetRole}
        mockedSetIsloggedIn={mockedSetIsloggedIn}
      />,
    );
    await waitFor(() => {
      expect(getJwtToken).toHaveBeenCalled();
      expect(mockedSetIsloggedIn).toHaveBeenCalled();
      expect(mockedGetRole).toHaveBeenCalled();
    });
    expect(asFragment()).toMatchSnapshot();
  });
  it("redirect to auth page", async () => {
    (getJwtToken as jest.Mock).mockRejectedValue("Mocked result");
    render(
      <WrapperComponent
        mockedGetRole={mockedGetRole}
        mockedSetIsloggedIn={mockedSetIsloggedIn}
      />,
    );
    await waitFor(() => {
      expect(getJwtToken).toHaveBeenCalled();
      expect(mockedSetIsloggedIn).not.toHaveBeenCalled();
      expect(mockedGetRole).not.toHaveBeenCalled();
    });
  });
});
