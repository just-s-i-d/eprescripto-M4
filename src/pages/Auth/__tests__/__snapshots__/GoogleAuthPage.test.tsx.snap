// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google Auth page tests Getting Jwt token and role 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center h-screen"
  >
    <img
      alt="loader"
      class="z-1 animate-loading"
      src="capsule.png"
      width="70"
    />
  </div>
</DocumentFragment>
`;
